src/main/java/org/extism/sdk/HostFunction.java

```diff
@@ -1,69 +1,72 @@
+    private final T userData;
 
-    public final Optional<T> userData;
+    public HostFunction(String name, LibExtism.ExtismValType[] params, LibExtism.ExtismValType[] returns, ExtismFunction<T> func) {
+        this(name, params, returns, func, null);
+    }
 
-    public HostFunction(String name, LibExtism.ExtismValType[] params, LibExtism.ExtismValType[] returns, ExtismFunction f, Optional<T> userData) {
+    public HostFunction(String name, LibExtism.ExtismValType[] params, LibExtism.ExtismValType[] returns, ExtismFunction<T> func, T userData) {
 
         this.name = name;
         this.params = params;
         this.returns = returns;
         this.userData = userData;
-        this.callback = (Pointer currentPlugin,
-                         LibExtism.ExtismVal inputs,
-                         int nInputs,
-                         LibExtism.ExtismVal outs,
-                         int nOutputs,
-                         Pointer data) -> {
-
-            LibExtism.ExtismVal[] outputs = (LibExtism.ExtismVal []) outs.toArray(nOutputs);
-
-            f.invoke(
-                    new ExtismCurrentPlugin(currentPlugin),
-                    (LibExtism.ExtismVal []) inputs.toArray(nInputs),
-                    outputs,
-                    userData
-            );
-
-            for (LibExtism.ExtismVal output : outputs) {
-                convertOutput(output, output);
-            }
-        };
+
+        int[] inputTypeValues = Arrays.stream(this.params).mapToInt(r -> r.v).toArray();
+        int[] outputTypeValues = Arrays.stream(this.returns).mapToInt(r -> r.v).toArray();
+        InternalExtismFunction callback = createCallbackFunction(func, userData);
+        Pointer userDataPointer = userData != null ? userData.getPointer() : null;
 
         this.pointer = LibExtism.INSTANCE.extism_function_new(
-                this.name,
-                Arrays.stream(this.params).mapToInt(r -> r.v).toArray(),
-                this.params.length,
-                Arrays.stream(this.returns).mapToInt(r -> r.v).toArray(),
-                this.returns.length,
-                this.callback,
-                userData.map(PointerType::getPointer).orElse(null),
+                name,
+                inputTypeValues,
+                inputTypeValues.length,
+                outputTypeValues,
+                outputTypeValues.length,
+                callback,
+                userDataPointer,
                 null
         );
     }
 
-    void convertOutput(LibExtism.ExtismVal original, LibExtism.ExtismVal fromHostFunction) {
-        if (fromHostFunction.t != original.t)
-            throw new ExtismException(String.format("Output type mismatch, got %d but expected %d", fromHostFunction.t, original.t));
+    private InternalExtismFunction createCallbackFunction(ExtismFunction<T> func, T userData) {
+        return (Pointer pluginPointer, ExtismVal inputs, int nInputs, ExtismVal outputs, int nOutputs, Pointer data) -> {
```

Comment:
the `data` pointer is not used here, is wrong?

Reply:
So, we probably don't need Extism to keep track of this pointer since we're using a closure and Java keeps the userData in scope. The pointer is for languages or functions that can't do this and they need to look back up the object from a pointer.

Reply:
Can we then leave out the data parameter on the interface?
If not, then we should document that the `data` parameter is not used here.

Reply:
I suspect we can leave it out, yes. If not then we should mark it as unused.

Reply:
The tests worked fine with this interface definition
```java
interface InternalExtismFunction extends Callback {

        /**
         * Host function implementation.
         */
        void invoke(Pointer plugin, ExtismVal inputs, int nInputs, ExtismVal outputs, int nOutputs
        //         , Pointer data
        );
    }
```

And this callback function:
```java
 private InternalExtismFunction createCallbackFunction(ExtismFunction<T> func, T userData) {
        return (Pointer pluginPointer, ExtismVal inputs, int nInputs, ExtismVal outputs, int nOutputs
        //         , Pointer data
        ) -> {

            var outputValues = (ExtismVal[]) outputs.toArray(nOutputs);
            var inputValues = (ExtismVal[]) inputs.toArray(nInputs);
            var currentPlugin = new ExtismCurrentPlugin(pluginPointer);

            func.invoke(currentPlugin, inputValues, outputValues, Optional.ofNullable(userData));

            for (ExtismVal output : outputValues) {
                coerceType(output);
            }
        };
    }
    ```